distributed/tests/test_nanny.py

```diff
@@ -781,3 +782,140 @@ class C:
+        self.in_instantiate = asyncio.Event()
+        self.wait_instantiate = asyncio.Event()
+
+    async def instantiate(self):
+        self.in_instantiate.set()
+        await self.wait_instantiate.wait()
+        return await super().instantiate()
+
+
+@pytest.mark.parametrize("restart", [True, False])
+@gen_cluster(client=True, nthreads=[])
+async def test_nanny_plugin_register_during_start_success(c, s, restart):
+    plugin = DummyNannyPlugin("foo", restart=restart)
+    n = SlowNanny(s.address)
+    assert not hasattr(n, "_plugin_registered")
+    start = asyncio.create_task(n.start())
+    try:
+        await n.in_instantiate.wait()
+
+        register = asyncio.create_task(c.register_worker_plugin(plugin))
+        with pytest.raises(asyncio.TimeoutError):
+            await asyncio.wait_for(asyncio.shield(register), timeout=0.1)
+        n.wait_instantiate.set()
+        assert await register
+        await start
+        assert n._plugin_registered
+    finally:
+        start.cancel()
+        await n.close()
+
+
+class SlowBrokenNanny(Nanny):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.in_instantiate = asyncio.Event()
+        self.wait_instantiate = asyncio.Event()
+
+    async def instantiate(self):
+        self.in_instantiate.set()
+        await self.wait_instantiate.wait()
+        raise RuntimeError("Nope")
+
+
+@pytest.mark.parametrize("restart", [True, False])
+@gen_cluster(client=True, nthreads=[])
+async def test_nanny_plugin_register_during_start_failure(c, s, restart):
+    plugin = DummyNannyPlugin("foo", restart=restart)
+    n = SlowBrokenNanny(s.address)
+    assert not hasattr(n, "_plugin_registered")
+    start = asyncio.create_task(n.start())
+    await n.in_instantiate.wait()
+
+    register = asyncio.create_task(c.register_worker_plugin(plugin))
+    with pytest.raises(asyncio.TimeoutError):
+        await asyncio.wait_for(asyncio.shield(register), timeout=0.1)
+    n.wait_instantiate.set()
+    with pytest.raises(RuntimeError):
+        await start
+    assert not await register
+
+
+class SlowDistNanny(Nanny):
+    def __init__(self, *args, in_instantiate, wait_instantiate, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.in_instantiate = in_instantiate
+        self.wait_instantiate = wait_instantiate
+
+    async def instantiate(self):
+        self.in_instantiate.set()
+        self.wait_instantiate.wait()
+        return await super().instantiate()
+
+
+def run_nanny(scheduler_addr, in_instantiate, wait_instantiate):
+    async def _():
+        worker = await SlowDistNanny(
+            scheduler_addr,
+            wait_instantiate=wait_instantiate,
+            in_instantiate=in_instantiate,
+        )
+        await worker.finished()
+
+    asyncio.run(_())
+
+
+@pytest.mark.parametrize("restart", [True, False])
+@gen_cluster(client=True, nthreads=[])
+async def test_nanny_plugin_register_nanny_killed(c, s, restart):
```

Comment:
@hendrikmakait this test is killing the nanny process while it is instantiating the Worker. That's probably the worst case because everything is in a partial state.

deadlocks with the latest commit but passes just fine now